### PR TITLE
docs: Document limitations.

### LINKED ISSUE #338: update limitations section for clear containers 3.0
Per discussion on https://github.com/clearcontainers/runtime/issues/12  :

Need to update known-limitations section to our project's readme.

### PR DESCRIPTION
**Pull request by jodh-intel**
Add a document listing the known limitations.

Fixes #338.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
Just noticed that the issue links aren't "active", but I think that's because the document isn't yet merged into the master branch.

**Comment by jodh-intel** [comment]
@iphutch - could you take a look at this please?

**Comment by grahamwhaley** [comment]
lgtm

File: docs/limitations.md [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+#### `docker run -m`
+
+The `docker run -m MEMORY` option is not currently supported. At the
+runtime level, this equates to the `linux.resources.memory` OCI
+configuration. It should be feasible to pass the relevant information
+through to the QEMU `-m` memory size option. This is also related to the
+`docker update` command.
+
+See issue #381 for more information.
```
Comment (gorozco1) [inline docs/limitations.md]: missed link to issue to make the document easy to read and to relate issues faster.

File: docs/limitations.md [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+#### `docker run --cpus=`
+
+The `docker run --cpus=` option is not currently implemented. At the
+runtime level, this equates to the `linux.resources.cpu` OCI
+configuration. It should be possible to pass this information through to
+the QEMU command line CPU configuration options to gain a similar
+effect.
+
+See issue #341 for more information.
```
Comment (gorozco1) [inline docs/limitations.md]: missed link to issue to make the document easy to read and to relate issues faster.

File: docs/limitations.md [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+#### `docker run --kernel-memory=`
+
+The `docker run --kernel-memory=` option is not currently implemented.
+It should be possible to pass this information through to the QEMU
+command line CPU configuration options to gain a similar effect.
+
+See issue #388 for more information.
```
Comment (gorozco1) [inline docs/limitations.md]: missed link to issue

File: docs/limitations.md [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+#### Capabilities
+
+The `docker run --cap-[add|drop]` commands are not supported by the
+runtime. At the runtime level, this equates to the
+`linux.process.capabilities` OCI configuration. Similar to the cgroup
+items, these capabilities could be modified either in the host, in the
+VM, or potentially both.
+
+See issue #51 for more information.
```
Comment (gorozco1) [inline docs/limitations.md]: missed link to issue

File: docs/limitations.md [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+The `docker stats` command does not return meaningful information for
+Clear Containers at present. This requires the runtime to support the
+`events` command.
+
+Some thought needs to go into if we display information purely from
+within the VM, or if the information relates to the resource usage of
+the whole VM container as viewed from the host. The latter is likely
+more useful from a whole system point of view.
+
+See issue #200 for more information.
```
Comment (gorozco1) [inline docs/limitations.md]: missed link to issue

File: docs/limitations.md [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+Note that the OCI standard does not specify a `stats` command.
+
+### runtime commands
+
+#### `ps` command
+
+The Clear Containers runtime does not currently support the `ps` command.
+
+See issue #95 for more information.
```
Comment (gorozco1) [inline docs/limitations.md]: missed link to issue

File: docs/limitations.md [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+The runtime does not currently implement the `events` command. We may
+not be able to perfectly match every sub-part of the `runc` events
+command, but we can probably do a subset, and maybe add some VM specific
+extensions.
+
+See here for the
+[runc implementation](https://github.com/opencontainers/runc/blob/e775f0fba3ea329b8b766451c892c41a3d49594d/events.go).
+
+See issue #379 for more information.
```
Comment (gorozco1) [inline docs/limitations.md]: missed link to issue

File: docs/limitations.md [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+`update` functionality is based around cgroup configuration.
+
+It may be possible to implement some of the update functionality by
+adjusting cgroups either around the VM or inside the container VM
+itself, or possibly by some other VM functional equivalent. It needs
+more investigation.
+
+Note that the OCI standard does not specify an `update` command.
+
+See issue #380 for more information.
```
Comment (gorozco1) [inline docs/limitations.md]: missed link to issue

**Comment by jodh-intel** [comment]
Hi @gorozco1 - gah! I had documented on this PR that the links should become active once it lands, but... you are right - they don't which is pretty sucky! ;-(

Changes:

- Made "internal" links active by specifying the full URL.
- Moved all "See issue ..." comments to the end of each section for consistency.
- Changed all `cc-oci-runtime` "See issue ..." comments to end with "... for more information." for consistency with the internal issue links.

**Comment by grahamwhaley** [comment]
Nice @jodh-intel . Still
lgtm
but you hit the dreaded CRI-O semaphore failure....

**Comment by grahamwhaley** [comment]
I could merge, but technically github would like @gorozco1 to OK the changes requested (broken links). let's wait for that ack...

**Comment by jodh-intel** [comment]
Wiki updated to point to the in-tree doc:

- https://github.com/clearcontainers/runtime/wiki/Current-Limitations

(and the draft wiki doc this PR was based upon has been deleted: https://github.com/clearcontainers/runtime/wiki/Current-Limitations-2).

**Pull request merged**